clients/api-d-roles/index.ts

```diff
@@ -0,0 +1,156 @@
+import { HttpServerError, HttpUnauthorizedError } from '#clients/exceptions';
+import routes from '#clients/routes';
+import constants from '#models/constants';
+import { Information } from '#models/exceptions';
+import httpClient, { httpGet, IDefaultRequestConfig } from '#utils/network';
+import { logInfoInSentry } from '#utils/sentry';
+import { URLSearchParams } from 'url';
+import {
+  IDRolesAuthTokenResponse,
+  IDRolesGroupSearchResponse,
+  IDRolesRoles,
+  IDRolesUser,
+} from './interface';
+
+/**
+ * D-Roles
+ * https://roles.preprod.data.gouv.fr/
+ */
+class DRolesAPIClient {
+  private _token: {
+    data: IDRolesAuthTokenResponse;
+    tokenExpiryTime: number;
+  } | null;
+
+  constructor(
+    private client_id: string | undefined,
+    private client_secret: string | undefined
+  ) {
+    if (
+      (!this.client_id || !this.client_secret) &&
+      process.env.NODE_ENV === 'production'
+    ) {
+      throw new HttpServerError('D-Roles env variables are undefined');
+    }
+    this._token = null;
+  }
+
+  private newToken = async () => {
+    try {
+      const data = await httpClient<IDRolesAuthTokenResponse>({
+        url: routes.dRoles.auth.token,
+        method: 'POST',
+        timeout: constants.timeout.XXXL,
+        headers: {
+          'Content-Type': 'application/x-www-form-urlencoded',
+        },
+        data: new URLSearchParams({
+          client_id: this.client_id || '',
+          client_secret: this.client_secret || '',
+          grant_type: 'client_credentials',
+        }).toString(),
+      });
+
+      this._token = {
+        data,
+        tokenExpiryTime: new Date().getTime() + 240 * 1000, // 4 minutes default expiry
+      };
+    } catch (e) {
+      this._token = null;
+      throw new HttpUnauthorizedError('Failed to get token');
+    }
+  };
+
+  private isTokenExpired = () => {
+    const now = new Date().getTime();
+    const tokenExpiryTime = this._token ? this._token.tokenExpiryTime : 0;
+    return now > tokenExpiryTime;
+  };
+
+  private getToken = async () => {
+    if (!this._token || this.isTokenExpired()) {
+      logInfoInSentry(
+        new Information({
+          name: 'RefreshingDRolesToken',
+          message: 'Refreshing D-Roles token',
+        })
+      );
+      await this.newToken();
+      if (!this._token) {
+        throw new HttpUnauthorizedError('Failed to refresh token');
+      }
+    }
+    return this._token;
+  };
+
+  private get = async <T>(
+    url: string,
+    config: IDefaultRequestConfig
+  ): Promise<T> => {
+    const token = await this.getToken();
+
+    return httpGet<T>(url, {
+      timeout: constants.timeout.M,
+      ...config,
+      headers: {
+        ...config.headers,
+        Authorization: `Bearer ${token.data.access_token}`,
+      },
+    });
+  };
```

Comment:
What happens when token is expired ? In InseeClient, I believe that token get reset to `null` thus ensuring you will re-run `getToken`

Also I think splitting client from the rest could improve readibility : 
```
client.tsx // auth + basic POST and GET wrapper
interface.tsx // types
index.tsx // all the routes wrapper
```

I advocate for this code splitting as it leave more room for core routes logic : 
- in case you want to do a mapToDomainObject, you can do it in `index.tsx`
- in case you want to have a file per group of routes you could even have `groups.ts, users.ts, roles.ts`

What you have written works great, as long as you dont use more routes or run heavier manipulation that just a basic `await`